server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -550,12 +551,19 @@ final SearchContext createAndPutContext(ShardSearchRequest request) throws IOExc
         SearchContext context = createContext(request);
         boolean success = false;
         try {
-            putContext(context);
             if (request.scroll() != null) {
                 openScrollContexts.incrementAndGet();
                 context.indexShard().getSearchOperationListener().onNewScrollContext(context);
             }
             context.indexShard().getSearchOperationListener().onNewContext(context);
+            putContext(context);
+            // ensure that if index is deleted concurrently, we free the context immediately, either here or in afterIndexRemoved
+            try {
+                indicesService.indexServiceSafe(request.shardId().getIndex());
+            } catch (IndexNotFoundException e) {
```

Comment:
I don't think you need the catch clause here - we free [below](https://github.com/elastic/elasticsearch/pull/42621/files#diff-a42a639093633662ea658bec29044a4cR571)

Reply:
How did I miss that, thanks! Fixed in 8edda7b36f7009f6d01d44fabe97e94153d7d06e

Reply:
but is this necessary or can we just rearrange the registration?